### PR TITLE
Features update logic for PPA in bionic

### PR DESCRIPTION
**Pull request by prathameshzarkar9**
[Issue 1062](https://github.com/devcontainers/features/issues/1062)
Features: Update logic for PPA in bionic

**Feature Name**
- common-utils
- go
- java
- node
- python

**Changelog**
- Changes to install.sh and devcontainer-feature.json for features update

**Checklist**
- [x] code snippet is added at specific position in all the install.sh of mentioned features
- [x] Bumped all the version for the mentioned features

### CODE REVIEW

File: src/java/install.sh [hunk 1]
```diff
@@ -53,6 +53,14 @@ else
     exit 1
 fi
 
+if [ "${ADJUSTED_ID}" = "rhel" ] && [ "${VERSION_CODENAME-}" = "centos7" ]; then
```
Comment (samruddhikhandale) [inline src/java/install.sh:56]: Let's resolve merge conflicts!

File: src/node/install.sh [hunk 2]
```diff
@@ -379,7 +387,7 @@ else
             [ ! -z "$http_proxy" ] && npm set proxy="$http_proxy"
             [ ! -z "$https_proxy" ] && npm set https-proxy="$https_proxy"
             [ ! -z "$no_proxy" ] && npm set noproxy="$no_proxy"
-            npm install -g pnpm
+            npm install -g @pnpm/exe
```
Comment (samruddhikhandale) [inline src/node/install.sh:390]: I think this is updated to fix the tests? However, both of these packages are different. 

Latest version of `pnpm` is not supported for node 16. Then, only for node 16, can we condition to install `pnpm 7` ? Then keep the rest as is! 

See https://pnpm.io/installation#compatibility

**Review by samruddhikhandale — CHANGES_REQUESTED** [review]

File: src/node/install.sh (hunk 2 above)
Comment (prathameshzarkar9) [reply src/node/install.sh]: Done

File: src/java/install.sh (hunk 1 above)
Comment (prathameshzarkar9) [reply src/java/install.sh]: resolved

File: src/node/install.sh [hunk 3]
```diff
@@ -370,20 +378,20 @@ if [ ! -z "${ADDITIONAL_VERSIONS}" ]; then
 fi
 
 # Install pnpm
-if bash -c ". '${NVM_DIR}/nvm.sh' && type pnpm >/dev/null 2>&1"; then
-    echo "pnpm already installed."
```
Comment (samruddhikhandale) [inline src/node/install.sh:374]: Why was this removed?

File: src/node/install.sh [hunk 4]
```diff
@@ -370,20 +378,20 @@ if [ ! -z "${ADDITIONAL_VERSIONS}" ]; then
...
+    (
+        . "${NVM_DIR}/nvm.sh"
+        [ ! -z "$http_proxy" ] && npm set proxy="$http_proxy"
+        [ ! -z "$https_proxy" ] && npm set https-proxy="$https_proxy"
+        [ ! -z "$no_proxy" ] && npm set noproxy="$no_proxy"
+        if [ "${NODE_VERSION}" = "16" ]; then
+            npm install -g --force pnpm@7
+        else
+            npm install -g --force pnpm
+        fi
```
Comment (samruddhikhandale) [inline src/node/install.sh:391]:
```
if [ "${NODE_VERSION}" = "16" ]; then
    PNPM_VERSION="pnpm@7"
else
    PNPM_VERSION="pnpm"
fi

npm install -g --force ${PNPM_VERSION}
```

Can we optimize?

**Review by samruddhikhandale — CHANGES_REQUESTED** [review]

File: src/node/install.sh (hunk 3 above)
Comment (prathameshzarkar9) [reply src/node/install.sh:374]: few images like `centos:centos7` has `pnpm` pre-installed which does not support the node version 16 so we should eliminate this condition and force install pnpm with the installed npm to support the compatible version of pnpm for node 16.

File: src/node/install.sh (hunk 4 above)
Comment (prathameshzarkar9) [reply src/node/install.sh:391]: done

File: src/node/install.sh (hunk 3 above)
Comment (samruddhikhandale) [reply src/node/install.sh:374]: Thanks for the thoughts here, however, we shouldn't override pnpm version if it already exists. This would have side effects on other distros which are neither centos nor are attempting to install node 16 

> few images like centos:centos7 has pnpm pre-installed which does not support the node version 16 

Did you find our tests failing?

File: src/node/install.sh [hunk 5]
```diff
@@ -370,20 +378,21 @@ if [ ! -z "${ADDITIONAL_VERSIONS}" ]; then
...
-if bash -c ". '${NVM_DIR}/nvm.sh' && type pnpm >/dev/null 2>&1"; then
-    echo "pnpm already installed."
+if bash -c ". '${NVM_DIR}/nvm.sh' && type npm >/dev/null 2>&1"; then
+    (
+        . "${NVM_DIR}/nvm.sh"
+        [ ! -z "$http_proxy" ] && npm set proxy="$http_proxy"
+        [ ! -z "$https_proxy" ] && npm set https-proxy="$https_proxy"
+        [ ! -z "$no_proxy" ] && npm set noproxy="$no_proxy"
+        PNPM_VERSION="pnpm"
+        if [ "${NODE_VERSION}" = "16" ]; then
```
Comment (samruddhikhandale) [inline src/node/install.sh:388]: Rethinking about this, in future it will end up piling too many conditions.

Is there a dynamic way to find out the compatibility between node versions and pnpm?

File: src/node/install.sh (hunk 3 above)
Comment (prathameshzarkar9) [reply src/node/install.sh:374]: kindly check below, if we use the default pnpm which is pre-installed in centos:7 scenaio. it is failing for pnpm -v check
```
The current version of Node.js is v16.20.2
Visit https://r.pnpm.io/comp to see the list of past pnpm versions with respective Node.js version support.


❌ pnpm check failed.
🧪 Executing duplicate test for feature 'node'...
⚠️ Skipping duplicate test for node because '/workspaces/features/test/node/duplicate.sh' does not exist.



  ================== TEST REPORT ==================
❌ Failed:      'centos-7'

@prathameshzarkar9 ➜ /workspaces/features (update-logic-for-ppa) $ docker exec -it 2f bash
[root@2f6f7f406396 /]# pnpm -v
ERROR: This version of pnpm requires at least Node.js v18.12
The current version of Node.js is v16.20.2
Visit https://r.pnpm.io/comp to see the list of past pnpm versions with respective Node.js version support.
[root@2f6f7f406396 /]# 
```

Comment (prathameshzarkar9) [reply src/node/install.sh:374]: @samruddhikhandale, reverting this change. let me know your thought on this.
adding a code to install pnpm specific version compatible for the installed node version

File: src/node/install.sh [hunk 6]
```diff
@@ -370,6 +378,32 @@ if [ ! -z "${ADDITIONAL_VERSIONS}" ]; then
 fi
 
 # Install pnpm
+# Declare an associative array
+declare -A node_pnpm_map
+
+# Populate the associative array with Node.js version to pnpm version mappings
+node_pnpm_map["10.13.0"]="5.18.10"
```
Comment (samruddhikhandale) [inline src/node/install.sh:385]: Thanks for adding it. I was hoping if there was a way to to dynamically find the compatible version without hardcoding, however, seems like there isn't a way.

Can we revert these changes? Let's add a new "pnpmVersion" Feature option where users can pin the `pnpm` version if they need to fix the compatibility issues.

Hence, in the test case which fails, we would use `pnpmVersion` to fix it.

**Review by samruddhikhandale — CHANGES_REQUESTED** [review]

File: src/node/install.sh (hunk 5 above)
Comment (prathameshzarkar9) [reply src/node/install.sh]: removed

File: src/common-utils/main.sh [hunk 7]
```diff
@@ -362,6 +362,14 @@ if [ "${ADJUSTED_ID}" = "rhel" ] && [ "${VERSION_CODENAME-}" = "centos7" ]; then
     sed -i s/^mirrorlist=http/#mirrorlist=http/g /etc/yum.repos.d/*.repo
 fi
 
+if [ "${ADJUSTED_ID}" = "rhel" ] && [ "${VERSION_CODENAME-}" = "centos7" ]; then
```
Comment (samruddhikhandale) [inline src/common-utils/main.sh:365]: Can we bump patch version? Looks like it got lost due to merging with main

File: src/java/devcontainer-feature.json [hunk 8]
```diff
@@ -1,6 +1,6 @@
 {
   "id": "java",
-  "version": "1.6.0",
+  "version": "1.6.1",
```
Comment (samruddhikhandale) [inline src/java/devcontainer-feature.json:3]: Why do we need to bump this given that we are not making any Java Feature changes?

File: src/node/devcontainer-feature.json [hunk 9]
```diff
@@ -28,6 +28,19 @@
...
+            "type": "string",
+            "proposals": [
+                "8.8.0",
+                "8.0.0",
+                "7.30.0",
+                "6.14.8",
+                "5.18.10",
+                "none"
+            ],
+            "default": "8.8.0",
```
Comment (samruddhikhandale) [inline src/node/devcontainer-feature.json:41]:
```suggestion
            "default": "latest",
```

Can you update to allow this ^ ?

File: src/node/devcontainer-feature.json [hunk 10]
```diff
@@ -28,6 +28,19 @@
...
+            "description": "Version of PNPM to install."
```
Comment (samruddhikhandale) [inline src/node/devcontainer-feature.json:42]:
```suggestion
            "description": "Select or enter the PNPM version to install"
```

File: src/node/install.sh [hunk 11]
```diff
@@ -8,6 +8,7 @@
 # Maintainer: The Dev Container spec maintainers
 
 export NODE_VERSION="${VERSION:-"lts"}"
+export PNPM_VERSION="${PNPMVERSION:-"none"}"
```
Comment (samruddhikhandale) [inline src/node/install.sh:11]:
```suggestion
export PNPM_VERSION="${PNPMVERSION:-"latest"}"
```

Comment (samruddhikhandale) [reply src/node/install.sh:11]: This should match with the defaults

File: src/node/install.sh [hunk 12]
```diff
@@ -370,7 +379,7 @@ if [ ! -z "${ADDITIONAL_VERSIONS}" ]; then
 fi
 
 # Install pnpm
-if bash -c ". '${NVM_DIR}/nvm.sh' && type pnpm >/dev/null 2>&1"; then
+if bash -c ". '${NVM_DIR}/nvm.sh' && [ ! -z "${PNPM_VERSION}" ] && [ "${PNPM_VERSION}" = "none" ]"; then
     echo "pnpm already installed."
```
Comment (samruddhikhandale) [inline src/node/install.sh:383]:
```suggestion
if [ ! -z "${PNPM_VERSION}" ] && [ "${PNPM_VERSION}" = "none" ]"; then
    echo "Ignoring installation of PNPM"
```

File: test/node/scenarios.json [hunk 13]
```diff
@@ -21,7 +22,8 @@
         "features": {
             "node": {
                 "version": "lts",
-                "additionalVersions": "v17.9.1,v14.19.3"
+                "additionalVersions": "v17.9.1,v14.19.3",
+                "pnpmVersion": "6.16.0"
```
Comment (samruddhikhandale) [inline test/node/scenarios.json:26]: Can we update the tests to verify the required version is installed?

**Review by samruddhikhandale — CHANGES_REQUESTED** [review]

File: src/common-utils/main.sh (hunk 7 above)
Comment (prathameshzarkar9) [reply src/common-utils/main.sh:365]: patch version bumped to "2.5.1"

File: src/node/devcontainer-feature.json (hunk 9 above)
Comment (prathameshzarkar9) [reply src/node/devcontainer-feature.json:41]: done

File: src/node/devcontainer-feature.json (hunk 10 above)
Comment (prathameshzarkar9) [reply src/node/devcontainer-feature.json:42]: done

File: src/node/install.sh (hunk 12 above)
Comment (prathameshzarkar9) [reply src/node/install.sh:383]: done

File: src/node/install.sh (hunk 11 above)
Comment (prathameshzarkar9) [reply src/node/install.sh:11]: done

File: test/node/scenarios.json (hunk 13 above)
Comment (prathameshzarkar9) [reply test/node/scenarios.json:26]: required version is verified in all the test cases for pnpmVersion

File: src/node/devcontainer-feature.json [hunk 14]
```diff
@@ -1,6 +1,6 @@
 {
     "id": "node",
-    "version": "1.5.0",
+    "version": "1.5.1",
```
Comment (samruddhikhandale) [inline src/node/devcontainer-feature.json:3]:
```suggestion
    "version": "1.6.0",
```

**Review by samruddhikhandale — APPROVED** [review]
Looks good!
